go/runtime/txpool/main_queue.go

```diff
@@ -0,0 +1,60 @@
+package txpool
+
+import (
+	"github.com/oasisprotocol/oasis-core/go/common/crypto/hash"
+	"github.com/oasisprotocol/oasis-core/go/common/logging"
+)
+
+var (
+	_ UsableTransactionSource     = (*mainQueue)(nil)
+	_ RecheckableTransactionStore = (*mainQueue)(nil)
+)
+
+// mainQueue is a priority queue for transactions that we give no special treatment.
+type mainQueue struct {
+	// This implementation adapts the existing scheduleQueue code.
+	inner scheduleQueue
+}
+
+func (mq *mainQueue) GetSchedulingSuggestion() [][]byte {
+	txMetas := mq.inner.getPrioritizedBatch(nil, 50)
+	var txs [][]byte
+	for _, txMeta := range txMetas {
+		txs = append(txs, txMeta.tx)
+	}
+	return txs
+}
+
+func (mq *mainQueue) GetTxByHash(h hash.Hash) ([]byte, bool) {
+	txMetas, _ := mq.inner.getKnownBatch([]hash.Hash{h})
+	if txMetas[0] == nil {
+		return nil, false
+	}
+	return txMetas[0].tx, true
+}
+
+func (mq *mainQueue) HandleTxsUsed(hashes []hash.Hash) {
+	mq.inner.remove(hashes)
+}
+
+func (mq *mainQueue) TakeAll() [][]byte {
+	txMetas := mq.inner.getAll()
+	mq.inner.clear()
+	var txs [][]byte
+	for _, txMeta := range txMetas {
+		txs = append(txs, txMeta.tx)
+	}
+	return txs
+}
+
+func (mq *mainQueue) OfferChecked(txs [][]byte) {
+	for _, tx := range txs {
+		txMeta := newTransaction(tx, txStatusChecked)
+		if err := mq.inner.add(txMeta); err != nil {
```

Comment:
I don't think this batch `OfferChecked` variant is ok as we need to be able to handle individual errors (at least in the current txpool implementation).

Reply:
ya the whole topic of how to notify the submitter of a transaction's fate is not settled yet. but for one thing, mainQueue is for txs from outside, and the submitter won't be some local listener we have. this will be more important in localQueue

Reply:
added error return